collector/extension/memorylimiterextension/memorylimiter.go

```diff
@@ -0,0 +1,165 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package memorylimiterextension // import "go.opentelemetry.io/collector/extension/memorylimiterextension"
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	arrowpb "github.com/open-telemetry/otel-arrow/api/experimental/arrow/v1"
+	"golang.org/x/sync/semaphore"
+	"go.uber.org/zap"
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+
+	"go.opentelemetry.io/collector/component"
+)
+
+type wrappedHandlerFn func(ctx context.Context, req any, handler grpc.UnaryHandler) (any, error)
+type memoryLimiterExtension struct {
+	limitBytes int64
+	sem        *semaphore.Weighted
+	logger     *zap.Logger
+	timeout    time.Duration
+	UnaryHandlerFn wrappedHandlerFn 
+}
+
+type requestWrappedServerStream struct {
+	wrappedCtx context.Context
+	stream grpc.ServerStream
+	pendingReqs chan any 
+	errCh chan error
+	timeout time.Duration
+}
+
+func (rss *requestWrappedServerStream) SetHeader(md metadata.MD) error {
+	return rss.stream.SetHeader(md)
+}
+
+func (rss *requestWrappedServerStream) SendHeader(md metadata.MD) error {
+	return rss.stream.SendHeader(md)
+}
+
+func (rss *requestWrappedServerStream) SetTrailer(md metadata.MD) {
+	rss.stream.SetTrailer(md)
+}
+
+func (rss *requestWrappedServerStream) Context() context.Context {
+	return rss.stream.Context()
+}
+
+func (rss *requestWrappedServerStream) SendMsg(m any) error {
+	return rss.stream.SendMsg(m)
+}
+
+func (rss *requestWrappedServerStream) RecvMsg(m any) error {
```

Comment:
This overrides the receivers call to Recv(). Since the underlying stream's RecvMsg() has already been called in the interceptor, we need to make sure the receiver still sees any pending request when it calls Recv(). e.g. for otelarrowreceiver this specific code fits in [Recv()](https://github.com/open-telemetry/otel-arrow/blob/main/collector/receiver/otelarrowreceiver/internal/arrow/arrow.go#L363) ->  [RecvMsg()](https://github.com/open-telemetry/otel-arrow/blob/main/api/experimental/arrow/v1/arrow_service_grpc.pb.go#L141)